src/tree.rs

```diff
@@ -1,22 +1,137 @@
 use lazy::Lazy;
+use std::fmt::{Debug, Error, Formatter};
+use std::rc::Rc;
 
+#[derive(Clone)]
 pub struct Tree<'a, A> {
     thunk: Lazy<'a, A>,
-    #[allow(dead_code)]
-    children: Vec<Tree<'a, A>>,
+    pub children: Vec<Tree<'a, A>>,
 }
 
-impl<'a, A: 'a + Clone> Tree<'a, A> {
+impl<'a, A> Tree<'a, A>
+where
+    A: 'a + Clone,
+{
+    fn new(value: A, children: Vec<Tree<'a, A>>) -> Self {
+        let thunk = Lazy::new(value);
+        Tree { thunk, children }
+    }
+
     pub fn singleton(value: A) -> Tree<'a, A> {
         Tree {
-            thunk: Lazy::new(move || value.clone()),
+            thunk: Lazy::new(value),
             children: vec![],
         }
     }
 
-    pub fn value(&mut self) -> &A {
+    pub fn value(&self) -> Option<A> {
```

Comment:
this isn't great as we always expect some value of A to come back, but it means a fair bit of play that's a tad distracting at the moment. I'll make an issue and come back to it later.